ext/cl_khr_defined_builtin_kernels.asciidoc

```diff
@@ -99,222 +103,414 @@ accelerators while being able to synchronize and optimize data transfers at
-a| -
+| *Kernel mode properties*
+| N/A
+| *Attribute validation rules*
 a|
-[start=0]
-. uchar* [out]: The data.
-* size_t* [in]: How many bytes to read before returning.
+* `shapeof(in) == shapeof(out)`
+* `elementof(in) == elementof(out)`
+* `alpha` must be a finite value.
+|===
+
+==== Launching DBKs from the Device Side ====
+
+DBKs are primarily meant to be launched as kernel commands via
+host-side command queues.  Optionally, they can be callable from
+device-side via `enqueue_kernel`:
+
+TBC. This probably needs device-side function corresponding to
+clCreateDefinedBuiltInKernelDescriptor.
+
+==== Sample Code ====
 
-4+| OpenCL C Semantics
-4+a|
 [source,c]
 ----
-__kernel void __khr_io_stream_in_blocking_uchar(uchar *output, size_t *num) {
-   while (*num) {
-       size_t num_read = *num;
-       __khr_io_stream_in_uchar(output, &num_read);
-       num -= num_read;
-       output += num_read;
-   }
+// TBD. Similarly in cl_qcom_ml_ops, tensors have type
+// (cl_channel_type) and a number of dimensions (rank) and dimension
+// sizes (shape). Difference over the cl_qcom_ml_ops is that the rank is
+// "unlimited".
+cl_tensor_desc_t lhs_tensor_desc = TBD;
+cl_tensor_desc_t rhs_tensor_desc = TBD;
+cl_tensor_desc_t res_tensor_desc = TBD;
+
+cl_dkb_attributes_matmul_v1 matmul_attrs = {
+  lhs_tensor_desc, rhs_tensor_desc, res_tensor_desc,
+  1, 0 // = Transpose lhs tensor
 }
-----
 
-|===
-
-==== Launching BiKs from the Device Side ====
-
-BiKs are primarily meant to be launched as kernel commands via host-side command queues.
-Optionally, they can be callable from device-side via
-`enqueue_kernel`: This capability can be queried on per BiK basis at compile-time in OpenCL C by checking for macro definitions which has the following naming convention: `cl_khr_bik_BUILTIN_KERNEL_NAME`. In case a BiK macro is defined, a kernel with a naming convention `__khr_BUILTIN_KERNEL_NAME()` can be enqueued by the program at device side as software-defined kernels.
+cl_dbk_mode_properties matmul_props = {
+  // Request a matmul implementation that meets this precision.
+  CL_DBK_PROPERTY_MAX_RELATIVE_ERROR, 100, // in ULPs.
+}
 
+cl_uint err;
+std::vector<cl_dbk_descriptor> kernel_descriptions;
+cl_dbk_descriptor matmul_desc =
+  clCreateDefinedBuiltInKernelDescriptor(
+  context, num_devices, device_list,
+  CL_DBK_MATMUL_V1, &matmul_attrs, &matmul_props, &err);
+
+} else if (err == CL_DBK_UNAVAILABLE) {
+  // Kernel attributes are valid but the kernel is not supported in at least
+  // one of the devices.
+} else if (err == CL_DBK_UNMET_MAX_RELATIVE_ERROR) {
+  // E.g. Kernel is supported but is not precise enough.
+} else if (err == CL_DBK_UNSUPPORTED_MODE_PROPERTY) {
+  // cl_dbk_mode_properties has a property not listed in the description of the
+  // defined built-in kernel.
+} else
+  kernel_descriptions.push_back(matmul_desc);
+
+...
+
+cl_program dbk_lib = clCreateProgramWithDefinedBuiltInKernels(
```

Comment:
OK. So the distinction to clCreateProgramWithBuiltInKernels is that we'd now communicate the relaxations in accuracy also here? Sounds fine to me. We can always create multiple programs if we need multiple accuracy levels for some reason.

Reply:
Not sure what you mean here. It is allowed to have two DBKs which are otherwise identical except for a mode property to live in the same program. IOW, one could build a program with matmul DBKs with different accuracy levels.